<a href="https://colab.research.google.com/github/bbrauser/DS-Unit-1-Sprint-1-Data-Wrangling-and-Storytelling/blob/master/BRAD_BRAUSER_LS_DS_121_Statistics_Probability_Assignment.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

<img align="left" src="https://lever-client-logos.s3.amazonaws.com/864372b1-534c-480e-acd5-9711f850815c-1524247202159.png" width=200>
<br></br>
<br></br>

## *Data Science Unit 1 Sprint 2 Assignment 1*

# Apply the t-test to real data

## Practice 1-Sample T-tests

One Sample t-tests determine whether or not a sample mean is statistically different from some known (or hypothesized) population mean. 

### 1) Load the Data
- Use the [automobile dataset](https://archive.ics.uci.edu/ml/datasets/Automobile)
- Fix the column headers
- Make sure NaNs are used to indicate missing values

Feel free to add code cells and text cells as needed throughout the assignment.

In [1]:
### YOUR WORK HERE
from google.colab import files
import pandas as pd
import numpy as np
from scipy import stats

uploaded = files.upload()

Saving imports-85.data to imports-85 (5).data


In [2]:
column_headers = ['Symboling', 'Normalized Losses', 'Make', 'Fuel Type', 'Aspiration', 
                  'Number of Doors', 'Body Style', 'Drive Wheels', 'Engine Location', 
                  'Wheel Base', 'Length', 'Width', 'Height', 'Curb Weight', 'Engine Type', 
                  'Number of Cylinders', 'Engine Size', 'Fuel System', 'Bore', 'Stroke', 
                  'Compression', 'Horsepower', 'Peak RPM', 'City MPG', 'Highway MPG', 'Price']

df = pd.read_csv('imports-85.data', header=None, names=column_headers)

df.head()

Symboling Normalized Losses         Make  ... City MPG Highway MPG  Price
0          3                 ?  alfa-romero  ...       21          27  13495
1          3                 ?  alfa-romero  ...       21          27  16500
2          1                 ?  alfa-romero  ...       19          26  16500
3          2               164         audi  ...       24          30  13950
4          2               164         audi  ...       18          22  17450

[5 rows x 26 columns]

In [3]:
df = df.replace({'?':np.NaN})

df.head()

Symboling Normalized Losses         Make  ... City MPG Highway MPG  Price
0          3               NaN  alfa-romero  ...       21          27  13495
1          3               NaN  alfa-romero  ...       21          27  16500
2          1               NaN  alfa-romero  ...       19          26  16500
3          2               164         audi  ...       24          30  13950
4          2               164         audi  ...       18          22  17450

[5 rows x 26 columns]

### 2) Pretend that this dataset represents the cars at a used car lot in your local town. 

- Use df.sample() to pick a random sample of 10 cars. Note that because this sample is random we are going to set the `random_state` so that all of us in the class get the same random sample. Please set your random state to `30` when using `df.sample()`

Is your sample reflective of the population value in regards to highway-mpg? Find the mean for `highway-mpg` for the entire dataset and compare your sample mean estimate to that value. Can you say that they are different? 

In [4]:
### YOUR WORK HERE

df_sample = df.sample(10, random_state=30)

mpg = df_sample['Highway MPG']

print(mpg.mean())
print(df['Highway MPG'].mean())

32.8
30.75121951219512


_Can you say that they are different?_

Yes, they are different.

The salesman says the cars he sells typically have a fuel efficiency of about 35 miles per gallon on the highway. You want to verify his claim but can only test 10 cars. Using your sample of 10, test his claim and report your results.

In [5]:
### YOUR WORK HERE

stats.ttest_1samp(mpg, 35)

Ttest_1sampResult(statistic=-0.9570244044334747, pvalue=0.36355472283248624)

From the sample data, With a t-statistic of 0.95 and a p-value of 0.363, I can conclude that the salesman's claim that the cars he sells have a fuel efficiency of about 35 miles is true.

Imagine that you now have the capacity to test 100 cars on the lot. Using the same random state of `30`, what would you conclude when sampling 100 cars?

In [6]:
### YOUR WORK HERE

df_sample_2 = df.sample(100, random_state=30)

mpg_2 = df_sample_2['Highway MPG']

stats.ttest_1samp(mpg_2, 35)

Ttest_1sampResult(statistic=-6.5185242467032305, pvalue=3.003137417468754e-09)

This 100-car sample, with a t-statistic of -6.52 and a p-value of 0, has shown that the salesman's claims of selling cars with around 35 highway mpg are actually false.

*Why might these two t-tests using the same dataset lead to different conclusions about the salesman's claim?*

The bigger the data size, the more precise the mean. A data set of only 10 cars produces mean that is not very refelctive of the population, but being able to test 100 cars will be substantially more representative.

## Practice 2-Sample T-tests

Two Sample t-tests determine whether or not two sample means are statistically different from each other. 

This portion of your assignment requires you to determine which issues have "statistically significant" differences between political parties in this [1980s congressional voting data](https://archive.ics.uci.edu/ml/datasets/Congressional+Voting+Records). 


### 1) Load the data

The data consists of 435 instances (one for each congressperson), a class (democrat or republican), and 16 binary attributes (yes or no for voting for or against certain issues). Be aware - there are missing values!

- Read the dataset in from UCI, you'll need to provide a list of column headers
- Encode "yes" votes as 1 and "no" votes as 0. (You can use `df.replace()` to swap out these values)
- Use dataframe filtering to split the dataframe into two new dataframes based on party. Hold all republicans in one dataframe and all democrats in the other. These will be our two different "samples."



In [7]:
### YOUR WORK HERE

uploaded2 = files.upload()

Saving house-votes-84.data to house-votes-84 (4).data


In [8]:
column_headers2 = ['Party', 'Handicapped Infants', 'Water Project Cost Sharing', 
                   'Adoption of the Budget Resolution', 'Physician Fee Freeze', 
                   'El Salvador Aid', 'Religious Groups in Schools', 
                   'Anti-Satellite Test Ban', 'Aid to Nicaraguan Contras', 
                   'Mix Missle', 'Immigration', 'Synfuels Corporation Cutback', 
                   'Education Spending', 'Superfun Right to Sue', 'Crime', 
                   'Duty-Free Exports', 'Export Administration Act South Africa']

df2 = pd.read_csv('house-votes-84.data', header=None, names=column_headers2)

df2 = df2.replace({'republican':'Republican', 'democrat':'Democrat', 'y':1, 'n':2, '?':np.NaN})

df2.head(10)

Party  ...  Export Administration Act South Africa
0  Republican  ...                                     1.0
1  Republican  ...                                     NaN
2    Democrat  ...                                     2.0
3    Democrat  ...                                     1.0
4    Democrat  ...                                     1.0
5    Democrat  ...                                     1.0
6    Democrat  ...                                     1.0
7  Republican  ...                                     1.0
8  Republican  ...                                     1.0
9    Democrat  ...                                     NaN

[10 rows x 17 columns]

In [10]:
dem_bool = df2['Party']=='Democrat'
rep_bool = df2['Party']=='Republican'

dem_vote = df2[dem_bool]
rep_vote = df2[rep_bool]

print(dem_vote.head())
print(rep_vote.head())

      Party  ...  Export Administration Act South Africa
2  Democrat  ...                                     2.0
3  Democrat  ...                                     1.0
4  Democrat  ...                                     1.0
5  Democrat  ...                                     1.0
6  Democrat  ...                                     1.0

[5 rows x 17 columns]
         Party  ...  Export Administration Act South Africa
0   Republican  ...                                     1.0
1   Republican  ...                                     NaN
7   Republican  ...                                     1.0
8   Republican  ...                                     1.0
10  Republican  ...                                     2.0

[5 rows x 17 columns]


In [11]:
dem_vote.head(20)

Party  ...  Export Administration Act South Africa
2   Democrat  ...                                     2.0
3   Democrat  ...                                     1.0
4   Democrat  ...                                     1.0
5   Democrat  ...                                     1.0
6   Democrat  ...                                     1.0
9   Democrat  ...                                     NaN
12  Democrat  ...                                     NaN
13  Democrat  ...                                     NaN
16  Democrat  ...                                     1.0
17  Democrat  ...                                     1.0
19  Democrat  ...                                     1.0
20  Democrat  ...                                     1.0
21  Democrat  ...                                     1.0
22  Democrat  ...                                     1.0
23  Democrat  ...                                     1.0
24  Democrat  ...                                     NaN
25  Democrat  ...                                     1.0
26  Democrat  ...                                     1.0
27  Democrat  ...                                     1.0
29  Democrat  ...                                     1.0

[20 rows x 17 columns]

In [12]:
rep_vote.head(20)

Party  ...  Export Administration Act South Africa
0   Republican  ...                                     1.0
1   Republican  ...                                     NaN
7   Republican  ...                                     1.0
8   Republican  ...                                     1.0
10  Republican  ...                                     2.0
11  Republican  ...                                     NaN
14  Republican  ...                                     NaN
15  Republican  ...                                     NaN
18  Republican  ...                                     2.0
28  Republican  ...                                     1.0
30  Republican  ...                                     2.0
33  Republican  ...                                     1.0
35  Republican  ...                                     2.0
36  Republican  ...                                     1.0
37  Republican  ...                                     1.0
38  Republican  ...                                     2.0
49  Republican  ...                                     2.0
51  Republican  ...                                     1.0
53  Republican  ...                                     2.0
55  Republican  ...                                     2.0

[20 rows x 17 columns]

### 2) Perform two sample T-tests on different issues and report the results.

- Find an issue that democrats support more than republicans with p < 0.01 (significant at the 99% level).
- Find an issue that republicans support more than democrats with p < 0.01 (significant at the 99% level).
- Find an issue where the difference between republicans and democrats has p > 0.1 (Not significant at the 90% level - i.e. there may not be much of a difference the two sample means)

Please for each test that you run state your null and alternative hypothesis and then write a conclusion reflecting on the null and alternative hypothesis.

Remember, that two-sample t-tests will only tell us if the two groups are *different* from one another. We'll have to look at their sample means directly or use the sign on the t-statistic to know which of the two sample means is larger. 

## An issue Democrats support more than Republicans - Adoption of the Budget Resolution

**Null Hypothesis:** Democrats support adopting the budget resolution more than Republicans.

**Alternative Hypothesis:** Democrats support adopting the budget resolution less than Republicans.

**Confidence Level:** 99%

In [25]:
stats.ttest_ind(dem_vote['Adoption of the Budget Resolution'], rep_vote['Adoption of the Budget Resolution'], nan_policy='omit')

Ttest_indResult(statistic=-23.212776917013784, pvalue=2.0703402795403286e-77)

**Conclusion:** With a t-statistic of -23.21 and a p-value of 0, the data suggests that Democrats supported adopting the budget resolution more than Republicans, which makes sense because the Democrats controlled the House in 1984.

## An issue Republican support more than Democrats - Crime

**Null Hypothesis:** Republicans support the crime bill more than Democrats

**Alternative Hypothesis:** Republicans support the crime bill less than Democrats

**Confidence Level:** 99%

In [27]:
stats.ttest_ind(dem_vote['Crime'], rep_vote['Crime'], nan_policy='omit')

Ttest_indResult(statistic=16.342085656197696, pvalue=9.952342705606092e-47)

**Conclusion:** With a t-statistic of 16.34 and a p-value of 0, the data would suggest that Republicans support the crime bill more than Democrats.

## An issue that seems to have bipartisan support - Water Project Cost Sharing

**Null Hypothesis:** Democrats and Republicans give roughly equal support to Water Project Cost Sharing.

**Alternative Hypothesis:** Democrats and Republicans give different levels of support. 

**Confidence Level:** 99%

In [28]:
stats.ttest_ind(dem_vote['Water Project Cost Sharing'], rep_vote['Water Project Cost Sharing'], nan_policy='omit')

Ttest_indResult(statistic=0.08896538137868286, pvalue=0.9291556823993485)

With a t-statistic of 0.09 and a p-value of 0.93, the Water Project Cost Sharing bill seems to have bipartisan support.

## Stretch Goals:

### 1) Use functions and some iterator (for loop, .apply(), list comprehension, etc.) to perform two sample t-tests on every issue in the dataset in an automated fashion.

In [38]:
### YOUR WORK HERE

for column in df2.columns[1:]:
  output.append(stats.ttest_ind(dem_vote[1:], rep_vote[1:], nan_policy='omit'))

TypeError: ignored



### 2) Work on Performing a T-test without using Scipy in order to get "under the hood" and learn more thoroughly about this topic.
### Start with a 1-sample t-test
 - Establish the conditions for your test 
 - [Calculate the T Statistic](https://blog.minitab.com/hs-fs/hubfs/Imported_Blog_Media/701f9c0efa98a38fb397f3c3ec459b66.png?width=247&height=172&name=701f9c0efa98a38fb397f3c3ec459b66.png) (You'll need to omit NaN values from your sample).
 - Translate that t-statistic into a P-value. You can use a [table](https://www.google.com/search?q=t+statistic+table) or the [University of Iowa Applet](https://homepage.divms.uiowa.edu/~mbognar/applets/t.html)

### Be sure to check your work using Scipy!





In [0]:
### YOUR WORK HERE

### 3) Then try a 2-sample t-test
 - Establish the conditions for your test 
 - [Calculate the T Statistic](https://lh3.googleusercontent.com/proxy/rJJ5ZOL9ZDvKOOeBihXoZDgfk7uv1YsRzSQ1Tc10RX-r2HrRpRLVqlE9CWX23csYQXcTniFwlBg3H-qR8MKJPBGnjwndqlhDX3JxoDE5Yg) (You'll need to omit NaN values from your sample).
 - Translate that t-statistic into a P-value. You can use a [table](https://www.google.com/search?q=t+statistic+table) or the [University of Iowa Applet](https://homepage.divms.uiowa.edu/~mbognar/applets/t.html)

 ### Be sure to check your work using Scipy!

In [0]:
### YOUR WORK HERE